pkg/daemon/update.go

```diff
@@ -2051,6 +2052,44 @@ func (dn *Daemon) experimentalUpdateLayeredConfig() error {
 			if err := client.RebaseLayered(desiredImage); err != nil {
 				return err
 			}
+
+			state, err := client.GetStatusStructured()
+			if err != nil {
+				return err
+			}
+			for _, deployment := range state.Deployments {
+				if deployment.Staged {
+					staged = deployment
+				}
+			}
+
+			// since we reboot whenever kargs change, we can get old kargs from live-applied or booted
+			oldContentBytes, err := Cat(booted.Osname, booted.Checksum, render.MCDContentPath)
+			if err != nil {
+				return err
+			}
+			newContentBytes, err := Cat(staged.Osname, staged.Checksum, render.MCDContentPath)
+			if err != nil {
+				return err
+			}
+			var oldContent, newContent render.MCDContent
+			if err := json.Unmarshal(oldContentBytes, &oldContent); err != nil {
+				return err
+			}
+			if err := json.Unmarshal(newContentBytes, &newContent); err != nil {
+				return err
+			}
+			// this shouldn't break transactionality because
+			// 1. if power is lost, rpm-ostree won't have changed the bootloader entry yet, so we'll reboot into the old image, and the rebase will be run again
+			// 2. we can't have graceful shutdown, because the MCO installs a SIGTERM handler
+			if err := dn.updateKernelArguments(oldContent.KernelArguments, newContent.KernelArguments); err != nil {
+				glog.Infof("Rolling back layered update since applying kernel arguments failed with: %w", err)
+				if cleanupErr := removePendingDeployment(); err != nil {
+					// at this point we have broken transactionality, but it's unlikely enough that we'll ignore for now
```

Comment:
@cgwalters is this comment correct? Do you agree?

Reply:
Yep, agreed.